src/app/utils/SLEF.js

```diff
@@ -0,0 +1,138 @@
+import {
+  CORE_INTERNAL_NAME_MAPPING,
+  CORIOLIS_TO_FD_BULKHEAD_NAME_MAPPING,
+  HARDPOINT_NUM_TO_CLASS,
+  SHIP_FD_NAME_TO_CORIOLIS_NAME,
+} from "./CompanionApiUtils.js";
+
+/**
+ * Converts the internal Coriolis representation to SLEF
+ * @param {typeof import('../shipyard/Ship.js')} ship Ship class
+ * @param {Object} internalData Internal JSON data
+ */
+export function toSLEF(ship, internalData) {
+  console.log({ ship });
+  const dataReference = internalData.references[0];
+  const fdShipName = Object.entries(SHIP_FD_NAME_TO_CORIOLIS_NAME).find(
+    ([_, coriolisName]) => coriolisName === ship.id,
+  )[0];
+
+  const slef = [
+    {
+      header: {
+        appName: dataReference.name ?? "Coriolis.io",
+        appVersion: "1.0",
+        appURL: dataReference.url ?? "",
+        appCustomProperties: {
+          ...dataReference,
+          shipName: internalData.name,
+        },
+      },
+      data: {
+        Ship: fdShipName,
+        HullValue: ship.m.discountedCost,
+        ModulesValue: ship.totalCost - ship.m.discountedCost,
+        UnladenMass: ship.unladenMass,
+        CargoCapacity: ship.cargoCapacity,
+        FuelCapacity: {
+          Main: ship.fuelCapacity,
+          Reserve: ship.reserveFuelCapacity,
+        },
+        Modules: [
+          {
+            Slot: CORE_INTERNAL_NAME_MAPPING.coriolisToFD.cargoHatch,
+            On: true,
+            Item: CORE_INTERNAL_NAME_MAPPING.fdToItemName.CargoHatch,
+            Priority: 0,
+            Value: 0,
+          },
+          {
+            Slot: CORE_INTERNAL_NAME_MAPPING.coriolisToFD.bulkheads,
+            On: true,
+            Item: `${fdShipName}_${CORIOLIS_TO_FD_BULKHEAD_NAME_MAPPING[ship.bulkheads.m.name]}`,
+            Priority: 0,
+            Value: ship.bulkheads.discountedCost,
+            Engineering: engineeringFromBlueprint(ship.bulkheads.m.blueprint),
+          },
+          ...ship.standard.map(coreInternalToSLEF),
+          ...ship.hardpoints.reduce(hardpointsToSLEF, []),
+          ...ship.internal.reduce(optionalInternalToSLEF, []),
```

Comment:
These could be maps, but they'd add the `undefined` values (which would be removed by the stringify) so, `reduce` already filters those out